transport/interceptor.go

```diff
@@ -0,0 +1,94 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package transport
+
+import "golang.org/x/net/context"
+
+// Interceptor defines a transport-level middleware for Inbounds.
+type Interceptor interface {
+	Apply(ctx context.Context, req *Request, resw ResponseWriter, handler Handler) error
```

Comment:
I can change the method names for now. Having a common `Middleware` type that can be used as either a filter or an interceptor is possible. We'll simply have to adapt that into a Filter/Interceptor.

If the common interface for Middleware is similar to the interface for Filter, there's an added cost of using that as an Interceptor because to return a `Response`, it'll have to take the full output of the `ResponseWriter` into a buffer.